### DISCUSSION TITLE
How to print out UInt8 strings directly from the serial port

### DISCUSSION THREAD

**Discussion by 654955321** [eModbus/eModbus#253] [Q&A] [answered]
First of all, thank you very much for providing such an excellent library. I have been able to read and write registers normally with your library and the examples provided. I am a newbie in arduino and I am having a problem to convert the read UInt16 string to UInt8.
As an example.
The string I am getting is
01 03 0A 56 2D 01 00 05 DC 00 00 00 00 00 16 F3
I want to print out :
“22061”-------corresponds to (56 2D)
“1”------------corresponds to (01 00)
“1500”-------- corresponds to (05 DC)

Maybe I am asking a low level question, because I have not touched C++ before, and I can't understand a big part of the library you provide, I hope I can get your help.

**Comment by Miq1** [comment, marked as answer]
Hi and welcome!
The data you are receiving is binary and the data formats will depend on the server you are requesting. If it is sending plain ``uint16-t`` values, these will consume 2 bytes each.
The easiest way to get these values from the response message is to use the ``get()`` function.
For example, if your response is in a variable named ``response``:
```
uint16_t offset = 3;  // the fourth byte in a FC03 response is the first byte of the data (counting from 0)
uint16_t a, b, c;  // three variables to hold data from the response
offset = response.get(offset, a);  // reads two bytes - as is the size of a - starting at offset into a, updating offset
offset = response.get(offset, b);  // same for b
offset = response.get(offset, c);  // same for c
```
I suppose you typed the response string in your post yourself? In the sense of Modbus it is inconsistent: the third byte is the length of the payload data following it. so normally I would expect 10 bytes in your example, but there are more? The ``01 00`` is not 1 but 256, by the way...

**Reply by 654955321** [reply]
**Thanks for your help, now with the code below I have it displayed properly.**

void handleData(ModbusMessage response, uint32_t token)
{
  uint16_t offset = 3;              // the fourth byte in a FC03 response is the first byte of the data (counting from 0)
  uint16_t a, b, c;                 // three variables to hold data from the response
  offset = response.get(offset, a); // reads two bytes - as is the size of a - starting at offset into a, updating offset
  offset = response.get(offset, b); // same for b
  offset = response.get(offset, c); // same for c
  float current = 0;
  current = (float)a / 1000;
  Serial.printf("Current:%0.2fA, Speed:%dRPM\r\n", current, b);
}
![image](https://user-images.githubusercontent.com/9605294/209064785-d4529d29-e504-4777-999f-1b4ee44eba43.png)


Next, I would like to ask you one more question. The experiment I am doing is using esp32 to communicate with the motor drive via modbus (RS485) to control it.
Controlled by the following code：

  err = MB.addRequest(Token++,1,WRITE_HOLD_REGISTER,0x0031,0x00C8);
  if (err != SUCCESS)
  {
    ModbusError e(err);
    Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
  }

"0x0031" is my register address (used to control the speed of the motor), I saw in your example that normally I only need to fill in "31" here, but I can only fill in "0x0031 " to access my register properly, what is the reason?

Also I want to make the written "0x00C8" a variable, and I want to change this variable through external buttons etc. to control the motor speed. Since I can only fill in the uint16 value here, I don't know how to implement it, so I hope to get your advice or help.

**Reply by Miq1** [reply]
The ``addRequest()`` family of calls is rather picky on the arguments you will give to it. Since the WRITE_HOLDING_REGISTER call is expecting both a register address and value of type ``uint_16t``, you can not provide anything else.
The ``0x0031`` is a hexadecimal representation of the decimal value 49. You should be able to use 49 or 0x31 as well. The value 0x00C8 of course can be given as a variable, but that needs to be of a type of ``uint16_t`` as well.

**Comment by 654955321** [comment]
Thank you very much for your patience in pointing me to understand more about the use of the library, the single motor drive communication control I have solved.

I am now trying to connect two motor drives. The problem I am having is that I can only read information from one drive. I have implemented the control of the motors separately via the pc uplink of the drive and have eliminated the wiring error. I think it is still a code problem and would like to get your guidance.

`#include <Arduino.h>
#include <SPI.h>
#include <Wire.h>
#include "ModbusClientRTU.h"
#include "ModbusMessage.h"

#define TXPIN GPIO_NUM_16   
#define RXPIN GPIO_NUM_15    

ModbusClientRTU MB(Serial2);
hw_timer_t *timer1 = NULL;


void handleData(ModbusMessage response, uint32_t token)
{
  uint16_t offset = 3;              // the fourth byte in a FC03 response is the first byte of the data (counting from 0)
  uint16_t a, b, c;                 // three variables to hold data from the response
  offset = response.get(offset, a); // reads two bytes - as is the size of a - starting at offset into a, updating offset
  offset = response.get(offset, b); // same for b
  offset = response.get(offset, c); // same for c
  float current = 0;
  current = (float)a / 100;
  Serial.printf("Current:%0.2fA, Speed:%dRPM\r\n", current, b);
}

void handleError(Error error, uint32_t token)
{
  ModbusError me(error);
  Serial.printf("Error response: %02X - %s\n", (int)me, (const char *)me);
}

void IRAM_ATTR onTimer1()
{
    uint32_t Token = 1111;
    Error err;

    err = MB.addRequest(Token++, 3, READ_HOLD_REGISTER, 0x0010, 3);
    if (err != SUCCESS)
    {
        ModbusError e(err);
        Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
    }

    err = MB.addRequest(Token++, 2, READ_HOLD_REGISTER, 0x0010, 3);
    if (err != SUCCESS)
    {
        ModbusError e(err);
        Serial.printf("Error creating request: %02X - %s\n", (int)e, (const char *)e);
    }
}
void setup()
{
  Serial2.begin(9600, SERIAL_8N1, TXPIN, RXPIN);
  MB.onDataHandler(&handleData);
  MB.onErrorHandler(&handleError);
  MB.setTimeout(2000);
  MB.begin();

  pinMode(1, INPUT);
  Serial.begin(9600);
  timer1 = timerBegin(1, 80, true);
  timerAttachInterrupt(timer1, &onTimer1, true);
  timerAlarmWrite(timer1, 2500000, true);

}

void loop()
{
  timerAlarmEnable(timer1);
}`

**Reply by 654955321** [reply]
This is the message of my serial port prompt：

![image](https://user-images.githubusercontent.com/9605294/209170003-f5593b46-bf1c-4cca-8bfc-af2c65e9aeaf.png)

This is the RS485 converter that I use：

![d68b9cc05f0285575789d11e45bbef0](https://user-images.githubusercontent.com/9605294/209170561-b502740d-4bf9-44c8-872d-83720e4ea554.jpg)

**Comment by Miq1** [comment]
You are getting timeout errors. That hints to serial communication with the second device is not working correctly. Are both devices connected daisy-chained or in a star topology?

**Reply by 654955321** [reply]
I wired it this way：
![image](https://user-images.githubusercontent.com/9605294/209257743-a9d89397-3da1-4297-b1e9-26f67623e472.png)

**Reply by Miq1** [reply]
That does show that the wiring in fact is connected star-wise, but in combination with the first picture of your setup I can see that the wires are short enough so that the topology will not matter.
I also doubt that a missing biasing circuit can cause the timeouts, else both motors would be affected evenly.
That somewhat is reducing the potential causes to the second motor and its RS485 connection. I suppose you will have checked the wiring several times to make sure everything is properly connected? Next to check would be the configuration: is it set to the right baud rate etc.? Is the server ID really set to 2 and 3?
Finally I would suggest trying another implementation. It is at least problematic to do too expensive things in a timer interrupt, since a couple of the ESP's mechanisms will be blocked while running it. My solution would be to only set a global flag in the interrupt routine and check that in ``loop()``. If it is set, then fire the two requests and reset it again.
As a bonus suggestion 😁 I would use a distinct and fixed ``Token`` value for each of the two requests and print that in the ``handle...`` callback functions. With it you better can tell which of the two motors is the one affected.

**Reply by 654955321** [reply]
Thanks for the reply; I eventually found out that it was still wired wrong. Reading the registers separately is now working fine, and I will continue to learn the usage of the library...
![image](https://user-images.githubusercontent.com/9605294/209365154-9fe5b3f4-0383-43de-ac07-7a83f2db17de.png)

**Reply by Miq1** [reply]
👍🏻

**Comment by 654955321** [comment]
Hello friends, I have another problem that I can't understand. About writing and reading negative numbers.

In the picture below the motor shows no problem reading the value in forward rotation, but in reverse rotation the value read should be negative, and I am reading a value that I cannot understand.

![image](https://user-images.githubusercontent.com/9605294/209818832-8f5180d7-ec37-4d0f-a594-4fece38104b4.png)

I then looked up the value of the motor in reverse when reading the register feedback as follows, and I converted it to a positive number, I could not figure out how it was converted to a negative number.

![image](https://user-images.githubusercontent.com/9605294/209819195-634ded27-b4e6-40b9-b066-b2a15545d3c9.png)

This is my code for reading the registers：

![image](https://user-images.githubusercontent.com/9605294/209819528-ce00064a-b1b4-4cba-9cdb-a09160a04bf7.png)

**Reply by Miq1** [reply]
Just use an ``int16_t`` variable instead of the ``uint16_t``.

**Reply by 654955321** [reply]
Thanks, it has been read successfully. But when I try to change the modified data type to write negative numbers, the library reports an error.

![image](https://user-images.githubusercontent.com/9605294/210067452-2f8bbc1b-22bb-4974-a4e1-c2fcb756708a.png)

![6BJO8D@6IY4%_M3VRON`P71](https://user-images.githubusercontent.com/9605294/210067483-5171f568-815e-47df-943f-197ad89b62da.png)

The error prompted：
   Error setMessage(uint8_t serverID, uint8_t functionCode, uint16_t p1, uint16_t p2, uint8_t count, uint16_t *arrayOfWords);
                                                                                                     ~~~~~~~~~~^~~~~~~~~~~~
*** [.pio\build\esp32-s3-devkitc-1\src\main.cpp.o] Error 1

**Reply by Miq1** [reply]
Yes, of course. The data in a Modbus message *per definitionem* is unsigned. You will need to use a cast for that, like ``(uint16_t)yourInt16Value``.

**Reply by 654955321** [reply]
Like this?
![image](https://user-images.githubusercontent.com/9605294/210068911-74ec690d-6b02-4192-8dfe-119ec73ed4a2.png)

**Reply by Miq1** [reply]
Yes.

**Reply by 654955321** [reply]
It doesn't seem to work, the value I want to write is "-1", but after I modified it the way you did, I found that the actual value written is "65535".

**Reply by Miq1** [reply]
Well, if taken as a signed ``int16_t`` value, 65535 **is** -1...
